src/main/java/com/iota/iri/utils/thread/BoundedScheduledExecutorService.java

```diff
@@ -0,0 +1,931 @@
+package com.iota.iri.utils.thread;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * This class represents a {@link SilentScheduledExecutorService} that accepts only a pre-defined amount of tasks that
+ * can be queued or executed at the same time. All tasks exceeding the defined limit will be ignored (instead of being
+ * queued) by either throwing a {@link RejectedExecutionException} or returning {@code null} depending on the method we
+ * call (non-silent vs silent).<br />
+ * <br />
+ * Whenever a non-recurring task finishes (or a recurring one is cancelled through its {@link Future}), it makes space
+ * for a new task. This is useful for classes like the {@link com.iota.iri.utils.log.interval.IntervalLogger} that want
+ * to delay an action if and only if there is no other delayed action queued already.<br />
+ * <br />
+ * Note: In contrast to other existing implementations like the SizedScheduledExecutorService of the apache package,
+ *       this class is thread-safe and will only allow to spawn and queue the exact amount of tasks defined during its
+ *       creation (since it does not rely on approximate numbers like the queue size).<br />
+ */
+public class BoundedScheduledExecutorService implements SilentScheduledExecutorService, ReportingExecutorService {
+    /**
+     * Holds the maximum amount of tasks that can be submitted for execution.<br />
+     */
+    private final int capacity;
+
+    /**
+     * Holds the underlying {@link ScheduledExecutorService} that manages the Threads in the background.<br />
+     */
+    private final ScheduledExecutorService delegate;
+
+    /**
+     * Holds a set of scheduled tasks tasks that are going to be executed by this
+     * {@link ScheduledExecutorService}.<br />
+     */
+    private final Set<TaskDetails> scheduledTasks = ConcurrentHashMap.newKeySet();
+
+    /**
+     * Thread-safe counter that is used to determine how many tasks were exactly scheduled already.<br />
+     * <br />
+     * Note: Whenever a task finishes, we clean up the used resources and make space for new tasks.<br />
+     */
+    private AtomicInteger tasksQueued = new AtomicInteger(0);
```

Comment:
Perhaps change the name to `tasksCounter` or `tasksQueuedCounter`

I got kinda of confused while reviewing

Reply:
renamed it